### PR TITLE
Update task version 1

### PR DESCRIPTION
**Pull request by yuege01**
# Implementation of UpdateTaskRun()
This pull request is to implement a UpdateTaskRun() function to receive update request from users and update taskruns 
in the database.
1. Currently, we use kubernetes UUID as key to update, we will use server generated results_id soon.
2. Due to lack of FieldMaskUtil API, right now we are updating whole taskrun instead of selected fields, may add this future in the future once we have the API

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as cyou
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by yuege01** [comment]
@wlynch @dibyom

File: results/cmd/api/main.go [hunk 1]
```diff
@@ -81,3 +81,27 @@ func (s *server) CreateTaskRun(ctx context.Context, req *pb.CreateTaskRunRequest
...
+		return nil, fmt.Errorf("failed to update a exsiting taskrun: %v", err)
+	}
+	taskrunFromClient := req.GetTaskRun()
+	blobData, err := proto.Marshal(taskrunFromClient)
+	if err != nil {
+		log.Println("taskrun marshaling error: ", err)
+		return nil, fmt.Errorf("taskrun marshaling error: %v", err)
+	}
+	taskrunMeta := taskrunFromClient.GetMetadata()
+	if _, err = statement.Exec(taskrunMeta.GetName(), taskrunMeta.GetNamespace(), blobData, taskrunMeta.GetUid()); err != nil {
```
Comment (wlynch) [inline results/cmd/api/main.go:102]:
```suggestion
	if _, err := statement.Exec(taskrunMeta.GetName(), taskrunMeta.GetNamespace(), blobData, taskrunMeta.GetUid()); err != nil {
```

File: results/cmd/api/main_test.go [hunk 2]
```diff
@@ -63,3 +64,27 @@ func setupTestDB(dbName string, t *testing.T) (*sql.DB, error) {
 	}
 	return db, nil
 }
+func TestUpdateTaskRun(t *testing.T) {
+	// Create a temporay database
```
Comment (wlynch) [inline results/cmd/api/main_test.go:68]:
```suggestion
	// Create a temporary database
```

File: results/demo/main.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	
+	
+			pods, err := clientset.CoreV1().Pods("").List(v1.ListOptions{})
+			if err != nil {
+				panic(err.Error())
+			}
+			for _, pod := range pods.Items {
+				fmt.Printf("%s %s\n", pod.GetName(), pod.GetCreationTimestamp())
+			}
```
Comment (wlynch) [inline results/demo/main.go:33]: Run this through gofmt.

File: results/demo/main.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+func main() {
+	kubeconfig := flag.String("kubeconfig", "/root/.kube/config", "absolute path to the kubeconfig file")
+	flag.Parse()
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		panic(err.Error())
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
```
Comment (wlynch) [inline results/demo/main.go:23]: I don't think we need to panic here. We can just `log.Fatal` the error

File: results/demo/main.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func main() {
+	kubeconfig := flag.String("kubeconfig", "/root/.kube/config", "absolute path to the kubeconfig file")
```
Comment (wlynch) [inline results/demo/main.go:15]: typically flags are defined in the package `var` section

Comment (wlynch) [inline results/demo/main.go:15]: We shouldn't recommend the use of root. We should default to the user's home directory.

Comment (wlynch) [reply results/demo/main.go:15]: e.g. `filepath.Join(os.Getenv("HOME"), ".kube", "config")`

File: results/demo/main.go (hunk 3 above)
Comment (wlynch) [inline results/demo/main.go:33]: This looks like this is just listing Pods on the cluster. Is this necessary? We may want to consider removing this file until we have something more meaningful to put here.

File: results/proto/api.proto [hunk 6]
```diff
@@ -20,6 +21,7 @@ message CreateTaskRunRequest {
 
 message UpdateTaskRunRequest {
 	TaskRun task_run = 1;
+	google.protobuf.FieldMask update_mask = 2;
```
Comment (wlynch) [inline results/proto/api.proto:24]: Proto files are often used as the API definition itself. It doesn't look like we're doing anything with this field, so we should remove this for the time being.

File: results/cmd/watcher/main.go [hunk 7]
```diff
@@ -54,25 +58,23 @@ func main() {
...
 		logger := logging.FromContext(ctx)
 		taskRunInformer := taskruninformer.Get(ctx)
-
 		c := &reconciler{
 			logger:        logger,
 			taskRunLister: taskRunInformer.Lister(),
 			client:        client,
 		}
-		impl := controller.NewImpl(c, c.logger, pipeline.PipelineRunControllerName)
+		impl := controller.NewImpl(c, c.logger, pipeline.TaskRunControllerName)
```
Comment (wlynch) [inline results/cmd/watcher/main.go:69]: These changes appear to be unrelated to adding support for `Update`. Let's break this out to it's own PR.

File: results/cmd/api/main.go [hunk 8]
```diff
@@ -81,3 +81,27 @@ func (s *server) CreateTaskRun(ctx context.Context, req *pb.CreateTaskRunRequest
...
+
+// UpdateTaskRun receives TaskRun and FieldMask from client and uses them to update records in local Sqlite Server.
+func (s *server) UpdateTaskRun(ctx context.Context, req *pb.UpdateTaskRunRequest) (*pb.TaskRun, error) {
+	database := s.db
+
+	// Update the entire row in database based on uid of taskrun.
+	statement, err := database.Prepare("UPDATE taskrun SET name = ?, namespace = ?, taskrunlog = ? WHERE uid = ?")
+	if err != nil {
+		log.Printf("failed to update a existing taskrun: %v\n", err)
+		return nil, fmt.Errorf("failed to update a exsiting taskrun: %v", err)
```
Comment (wlynch) [inline results/cmd/api/main.go:93]:
```suggestion
		return nil, fmt.Errorf("failed to update a exsiting taskrun: %w", err)
```

`%w` is generally preferred over `%v` when returning errors since it preserves underlying error information. See https://blog.golang.org/go1.13-errors for more details.

File: results/cmd/api/main.go [hunk 9]
```diff
@@ -81,3 +81,27 @@ func (s *server) CreateTaskRun(ctx context.Context, req *pb.CreateTaskRunRequest
...
+	}
+	taskrunFromClient := req.GetTaskRun()
+	blobData, err := proto.Marshal(taskrunFromClient)
+	if err != nil {
+		log.Println("taskrun marshaling error: ", err)
+		return nil, fmt.Errorf("taskrun marshaling error: %v", err)
```
Comment (wlynch) [inline results/cmd/api/main.go:99]:
```suggestion
		return nil, fmt.Errorf("taskrun marshaling error: %w", err)
```

File: results/cmd/api/main_test.go [hunk 10]
```diff
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	pb "github.com/tektoncd/experimental/results/proto/proto"
+	fieldmask "google.golang.org/genproto/protobuf/field_mask"
```
Comment (wlynch) [inline results/cmd/api/main_test.go:12]: This library is unused, let's remove it. (Running `goimports` should fix this automatically!)

File: results/cmd/api/main.go [hunk 11]
```diff
@@ -81,3 +81,27 @@ func (s *server) CreateTaskRun(ctx context.Context, req *pb.CreateTaskRunRequest
 	}
 	return taskrunFromClient, nil
 }
+
+// UpdateTaskRun receives TaskRun and FieldMask from client and uses them to update records in local Sqlite Server.
+func (s *server) UpdateTaskRun(ctx context.Context, req *pb.UpdateTaskRunRequest) (*pb.TaskRun, error) {
+	database := s.db
```
Comment (wlynch) [inline results/cmd/api/main.go:87]: nit: It looks like this is only used once, so I'm not sure if this is strictly necessary (e.g. we could just use `s.db.Prepare` below instead). What do you think?

File: results/cmd/api/main.go [hunk 12]
```diff
@@ -81,3 +81,27 @@ func (s *server) CreateTaskRun(ctx context.Context, req *pb.CreateTaskRunRequest
...
+	taskrunFromClient := req.GetTaskRun()
+	blobData, err := proto.Marshal(taskrunFromClient)
+	if err != nil {
+		log.Println("taskrun marshaling error: ", err)
+		return nil, fmt.Errorf("taskrun marshaling error: %v", err)
+	}
+	taskrunMeta := taskrunFromClient.GetMetadata()
+	if _, err = statement.Exec(taskrunMeta.GetName(), taskrunMeta.GetNamespace(), blobData, taskrunMeta.GetUid()); err != nil {
+		log.Printf("failed to execute update of a new taskrun: %v\n", err)
+		return nil, fmt.Errorf("failed to execute update of a new taskrun: %v", err)
```
Comment (wlynch) [inline results/cmd/api/main.go:104]:
```suggestion
		return nil, fmt.Errorf("failed to execute update of a new taskrun: %w", err)
```

File: results/demo/main.go (hunk 3 above)
Comment (yuege01) [reply results/demo/main.go:33]: oh, later I found this one is useless I just used it for debugging watcher. Will delete it.

File: results/demo/main.go (hunk 4 above)
Comment (yuege01) [reply results/demo/main.go:23]: Will delete it.

File: results/demo/main.go (hunk 5 above)
Comment (yuege01) [reply results/demo/main.go:15]: will delete it.

File: results/cmd/api/main_test.go (hunk 10 above)
Comment (yuege01) [reply results/cmd/api/main_test.go:12]: It is actually used in TestUpdatedTaskRun to create a fieldmask

File: results/cmd/api/main.go (hunk 1 above)
Comment (yuege01) [reply results/cmd/api/main.go]: Done

File: results/cmd/api/main_test.go (hunk 2 above)
Comment (yuege01) [reply results/cmd/api/main_test.go]: Done

File: results/demo/main.go (hunk 3 above)
Comment (yuege01) [reply results/demo/main.go]: Removed the file since it is unused now.

File: results/demo/main.go (hunk 4 above)
Comment (yuege01) [reply results/demo/main.go]: done

File: results/demo/main.go (hunk 5 above)
Comment (yuege01) [reply results/demo/main.go]: done

Comment (yuege01) [reply results/demo/main.go]: deleted it. but good advice, thanks!

File: results/demo/main.go (hunk 3 above)
Comment (yuege01) [reply results/demo/main.go]: done

File: results/cmd/watcher/main.go (hunk 7 above)
Comment (yuege01) [reply results/cmd/watcher/main.go]: Sure

File: results/cmd/api/main.go (hunk 8 above)
Comment (yuege01) [reply results/cmd/api/main.go]: Thanks for the advice

Comment (yuege01) [reply results/cmd/api/main.go]: done

File: results/cmd/api/main.go (hunk 9 above)
Comment (yuege01) [reply results/cmd/api/main.go]: done

File: results/cmd/api/main.go (hunk 11 above)
Comment (yuege01) [reply results/cmd/api/main.go]: done, I have deleted this line

File: results/cmd/api/main.go (hunk 12 above)
Comment (yuege01) [reply results/cmd/api/main.go]: done

File: results/proto/api.proto (hunk 6 above)
Comment (yuege01) [reply results/proto/api.proto:24]: Do you mean it is okay for now not to follow api definition in Google AIP guideline?

File: results/cmd/api/main_test.go (hunk 10 above)
Comment (wlynch) [reply results/cmd/api/main_test.go:12]: Looking at the implementation of Update, it doesn't look like this is used. =\

This actually points out something important though! We should probably validate the result of the update somehow, either by using Get or reading the DB.

File: results/cmd/watcher/main.go [hunk 13]
```diff
@@ -83,6 +85,7 @@ type reconciler struct {
 
 func (r *reconciler) Reconcile(ctx context.Context, key string) error {
 	r.logger.Infof("reconciling resource key: %s", key)
+	print(ctx)
```
Comment (wlynch) [inline results/cmd/watcher/main.go:88]: Avoid using the builtin print. You should be able to use r.logger instead.

File: results/proto/api.proto (hunk 6 above)
Comment (wlynch) [reply results/proto/api.proto:24]: Discussed offline- the ultimate goal is to support this, but for now it doesn't make much sense to have a field that isn't actually used.

File: results/cmd/api/main_test.go (hunk 10 above)
Comment (yuege01) [reply results/cmd/api/main_test.go:12]: oh, you mean in the UpdateTaskRun function itself. I will delete it along with its definition in proto file.

File: results/demo/main.go (hunk 3 above)
Comment (yuege01) [reply results/demo/main.go]: done

File: results/cmd/api/main_test.go (hunk 10 above)
Comment (yuege01) [reply results/cmd/api/main_test.go]: Done

File: results/cmd/watcher/main.go (hunk 13 above)
Comment (yuege01) [reply results/cmd/watcher/main.go]: Just deleted it.

File: results/proto/api.proto (hunk 6 above)
Comment (yuege01) [reply results/proto/api.proto]: Done

File: results/cmd/api/main_test.go (hunk 10 above)
Comment (yuege01) [reply results/cmd/api/main_test.go]: I also added a further test to see if values really have changed.

File: results/cmd/api/main_test.go [hunk 14]
```diff
@@ -61,5 +65,46 @@ func setupTestDB(dbName string, t *testing.T) (*sql.DB, error) {
...
+	// Validate by checking if we can get the updated taskrun
+	rows, err := srv.db.Query("SELECT taskrunlog FROM taskrun WHERE uid = ?", r.GetMetadata().GetUid())
+	if err != nil {
+		t.Fatalf("failed to query on database: %v", err)
+	}
+	for rows.Next() {
+		var taskrunblob []byte
+		taskrun := &pb.TaskRun{}
+		rows.Scan(&taskrunblob)
+		if err = proto.Unmarshal(taskrunblob, taskrun); err != nil {
```
Comment (wlynch) [inline results/cmd/api/main_test.go:102]:
```suggestion
		if err := proto.Unmarshal(taskrunblob, taskrun); err != nil {
```

File: results/cmd/watcher/main.go [hunk 15]
```diff
@@ -70,9 +72,9 @@ func main() {
 			AddFunc:    impl.EnqueueControllerOf,
 			UpdateFunc: controller.PassNew(impl.EnqueueControllerOf),
 		})
-
 		return impl
 	})
+	println()
```
Comment (wlynch) [inline results/cmd/watcher/main.go:77]: Remove

File: results/cmd/watcher/main.go [hunk 16]
```diff
@@ -24,19 +24,23 @@ import (
...
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+
 	taskruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1beta1/taskrun"
 	listers "github.com/tektoncd/pipeline/pkg/client/listers/pipeline/v1beta1"
+
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Comment (wlynch) [inline results/cmd/watcher/main.go:35]: What is this used for? (as well as knative.dev/pkg/system/testing)

File: results/cmd/watcher/main.go [hunk 17]
```diff
@@ -24,19 +24,23 @@ import (
 	"github.com/tektoncd/experimental/results/pkg/convert"
 	pb "github.com/tektoncd/experimental/results/proto/proto"
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+
```
Comment (wlynch) [inline results/cmd/watcher/main.go:27]: nit: Don't separate these from the rest of the tekton packages.

File: results/cmd/api/main_test.go [hunk 18]
```diff
@@ -61,5 +65,46 @@ func setupTestDB(dbName string, t *testing.T) (*sql.DB, error) {
...
+			t.Fatal("unmarshaling error: ", err)
+		}
+		metadata := taskrun.GetMetadata()
+		if metadata.GetName() != r.GetMetadata().GetName() || metadata.GetNamespace() != r.GetMetadata().GetNamespace() {
+			t.Fatal("Update Function not properly implemented.")
+		}
```
Comment (wlynch) [inline results/cmd/api/main_test.go:108]: Use https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc#Diff instead to compare the entire struct.

File: results/cmd/watcher/main.go (hunk 16 above)
Comment (yuege01) [reply results/cmd/watcher/main.go:35]: Without this import, when I ran the watcher.go, in gave me an error

File: results/cmd/api/main_test.go (hunk 14 above)
Comment (yuege01) [reply results/cmd/api/main_test.go]: Done

File: results/cmd/watcher/main.go (hunk 15 above)
Comment (yuege01) [reply results/cmd/watcher/main.go]: Done

File: results/cmd/watcher/main.go (hunk 16 above)
Comment (yuege01) [reply results/cmd/watcher/main.go:35]: Done

File: results/cmd/api/main_test.go (hunk 18 above)
Comment (yuege01) [reply results/cmd/api/main_test.go]: Done

File: results/cmd/watcher/main.go (hunk 17 above)
Comment (yuege01) [reply results/cmd/watcher/main.go]: Done

**Comment by wlynch** [comment]
Mostly LGTM! Just one thing I think we can do to simplify. Once the Get change is in, let's make the changes and squash, as well as update the commit/PR description.

File: results/go.sum [hunk 19]
```diff
@@ -484,7 +492,20 @@ github.com/stretchr/testify v1.3.0/go.mod h1:M5WIy9Dh21IEIfnGCwXGc5bZfKNJtfHm1UV
 github.com/stretchr/testify v1.4.0/go.mod h1:j7eGeouHqKxXV5pUuKE4zz7dFj8WfuZ+81PSLYec5m4=
 github.com/stretchr/testify v1.5.1 h1:nOGnQDM7FYENwehXlg/kFVnos3rEvtKTjRvOWSzb6H4=
 github.com/stretchr/testify v1.5.1/go.mod h1:5W2xD1RspED5o8YsWQXVCued0rvSQ+mT+I5cxcmMvtA=
+<<<<<<< HEAD
```
Comment (wlynch) [inline results/go.sum:495]: Fix diffs

File: results/cmd/watcher/main.go [hunk 20]
```diff
@@ -26,17 +26,22 @@ import (
...
 	"k8s.io/client-go/tools/cache"
 	"knative.dev/pkg/configmap"
 	"knative.dev/pkg/controller"
 	"knative.dev/pkg/injection"
 	"knative.dev/pkg/injection/sharedmain"
 	"knative.dev/pkg/logging"
 	"knative.dev/pkg/signals"
+
+	// this import ensures watcher connect to cluster
+	_ "knative.dev/pkg/system/testing"
```
[the remark this answers is not in the document]
Comment (yuege01) [reply results/cmd/watcher/main.go]: done

File: results/go.sum (hunk 19 above)
Comment (yuege01) [reply results/go.sum]: done
